src/controller.coffee

```diff
@@ -1179,65 +1184,95 @@ define ['droplet-helper',
   hook 'mouseup', 1, (point, event, state) ->
     # We will consume this event iff we dropped it successfully
     # in the root tree.
-    if @draggingBlock? and @lastHighlight?
-      if @inTree @draggingBlock
-        # Since we removed this from the tree,
-        # we will need to log an undo operation
-        # to put it back in.
-        @addMicroUndoOperation 'CAPTURE_POINT'
+    if @draggingBlock?
+      if not @currentlyUsingBlocks
+        # See if we can drop the block's text in ace mode.
+        position = new @draw.Point(
+          point.x + @draggingOffset.x,
+          point.y + @draggingOffset.y
+        )
+
+        if @trackerPointIsInAce position
+          # Get the line of text we're dropping into
+          pos = @aceEditor.renderer.screenToTextCoordinates position.x, position.y
+          line = @aceEditor.session.getLine pos.row
+          leadingWhitespace = /^(\s*)/.exec(line)[0]
+          prefix = ''
+          if pos.column == line.length and leadingWhitespace.length != line.length
+            # We're at the end of a non-empty line.
+            # Stick our inserted text on a new line
+            # with the same indentation
+            prefix = '\n' + leadingWhitespace
+
+          # Call prepareNode, which may append with a semicolon
+          @prepareNode @draggingBlock, null
+          text = prefix + @draggingBlock.stringify @mode
+
+          if not prefix and text[text.length - 1] == ';'
+            # Add a potentially indented new line if
+            # we didn't already do so.
+            text += '\n' + leadingWhitespace
+          @aceEditor.onTextInput text
```

Comment:
Not sure if all of this logic around adding newlines/indentation makes more sense existing elsewhere.